numeric_types/half.h

```diff
@@ -48,8 +48,60 @@ struct half {
     data = halfFromFloat(value);
     return *this;
   }
+
 };
 
+inline half operator + (const half& a, const half& b) {
```

Comment:
Although the expectation is that host code is ahead-of-time and not performance critical, if we're going to include these at all and it's not *too* complex to implement something reasonably performant we should consider that. We're already citing Mike Acton's implementation for the conversions, there's at least some reference implementations here https://github.com/macton/ps3-archive/blob/master/src/half-precision/Half.v2/int_insn.h 

Alternatively we could leave this out of scope unless we have a strong case for this being valuable. It could be a reasonable expectation for users to bring their own computations on the host side - they could use f32 or their own choice of library or implementation and limit gpu.cpp to casting / data movement.

Maybe there's a rationale that's not obvious, but it seems like having operator implementations that are implemented via casts internally could be worse than either option? If f32 conversions are happening within operations, wouldn't it be better to use f32 and cast at the end?

Reply:
randn needs these operators  to generate values of half type.
But I can cast float32 values to half values after processing randn.